### ISSUE TITLE
TLSv1.3 Exporters

### ISSUE THREAD

**Issue by tatsuhiro-t** [openssl/openssl#3680]
https://tools.ietf.org/html/draft-ietf-tls-tls13-20#section-7.5

TLSv1.3 defines exporters which use different construction from the original RFC 5705.
OpenSSL has  SSL_export_keying_material, but it seems that , even if TLSv1.3 is negotiated, the underlying export function is still tls1_export_keying_material, which uses PRF version.

**Comment by mattcaswell** [comment]
Yeah, this is a known issue. Well, it's known to me anyway. It's been on my TODO list for a while. I'm hoping to get to it "soon". Probably after the PSK work (#3670) is complete.

**Comment by tatsuhiro-t** [comment]
Thank you!

I tested this exporter after SSL_do_handshake returns 1.
But it looks like derived keys disagree between client and server.

**Comment by mattcaswell** [comment]
> But it looks like derived keys disagree between client and server.

Hmm. Can you share some code? The test I added checks exactly that, i.e. generate the same keys on both client and server and confirm that they are the same.

**Comment by tatsuhiro-t** [comment]
Not something special, but here they are:

```cpp
int export_secret(uint8_t *dest, size_t destlen, SSL *ssl, const uint8_t *label,
                  size_t labellen) {
  int rv;

  rv = SSL_export_keying_material(
      ssl, dest, destlen, reinterpret_cast<const char *>(label), labellen,
      reinterpret_cast<const uint8_t *>(""), 0, 1);
  if (rv != 1) {
    return -1;
  }

  return 0;
}

int export_client_secret(uint8_t *dest, size_t destlen, SSL *ssl) {
  constexpr uint8_t label[] = "EXPORTER-QUIC client 1-RTT Secret";
  return export_secret(dest, destlen, ssl, label, str_size(label));
}

int export_server_secret(uint8_t *dest, size_t destlen, SSL *ssl) {
  constexpr uint8_t label[] = "EXPORTER-QUIC server 1-RTT Secret";
  return export_secret(dest, destlen, ssl, label, str_size(label));
}
```

**Comment by tatsuhiro-t** [comment]
BTW, this works with boringssl...

**Comment by mattcaswell** [comment]
So, just to be clear, you are extracting two secrets on both sides? i.e. you call export_client_secret() on both the client and the server and they are different? Similarly with export_server_secret()?

**Comment by tatsuhiro-t** [comment]
Yes. They are used for sending and receiving data respectively.
For client, client secret is used to encrypt data, and server secret is
used to decrypt data from server.
For server, server secret is used to encrypt data, and client secret is
used to decrypt data from client.


2017/06/30 0:02 "mattcaswell" <notifications@github.com>:

So, just to be clear, you are extracting two secrets on both sides? i.e.
you call export_client_secret() on both the client and the server and they
are different? Similarly with export_server_secret()?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/openssl/openssl/issues/3680#issuecomment-311994529>,
or mute
the thread

**Comment by tatsuhiro-t** [comment]
https://github.com/openssl/openssl/blob/2915fe19a6676374c335d8c50eaaa4c940cf47d6/ssl/tls13_enc.c#L625
The last argument is suspicious.  I think it should be hashsize, otherwise we get random garbage in exportersecret.  If I did it, I got the same key from both client and server.

**Comment by mattcaswell** [comment]
Good catch. Fix in #3810.

**Comment by mattcaswell** [comment]
Should be fixed now.
